### PR TITLE
[DNM] Distributed fixes

### PR DESCRIPTION
**Pull request by Islandman93**
This changes update distributed to support tensorflow summaries, fixes some important issues,  and allows q_model target networks to be updated in distributed mode.

## Tensorflow Summaries
The "chief" agent (task index 0) will now output summaries in distributed mode. I had to add a supervisor argument into the set_session method on model.py because summaries can only be written through the supervisor in distributed mode. I've modified the distributed example to remove summaries from all workers except task 0.

## Distributed Fixes

1. Global shared optimizer, according to the A3C paper using a shared optimizer produces better results than each worker having it's own. This might make workers like 0.001% and save some memory since they don't have to calculate the optimizer update or store the params just the gradients.
2. Updating local vars. Fixed local vars were being updated from a cached copy of the global variables. Specifically, var.read_value() must be used to ensure tensorflow doesn't used the cached copy. Additionally, local vars are only updated after the gradient calculation. Reset now causes a copy from global -> local, this fixes problems where workers that start slowly had really old parameters. Or if a worker needs to be restarted it will copy from global first.
3. Added general helper methods that allow models to override how the global_timestep is incremented. This allows say DQN to only increment the timestep once even if batch size is 32. 

## Q Model target
Was able to get a tensorflow conditional statement that determines if the global model needs to update the target network. This was the major issue preventing distributed running of DQN & variants. So now all DQN methods should work in a distributed fashion.

I've tested Nstep DQN with these changes and it's now capable of running for an epoch or two before gradient explosion (up from about 10k steps to 1.2mil). I think that this is just a hyperparameter issue and that the distributed code is now correct (NStep DQN still doesn't clip gradients like the original paper so one bad update can mess up the whole thing). 

The policy gradient tests are failing because they don't have optimizers? If that's expected behavior then I can remove the check from the code.

### CODE REVIEW

**Comment by Islandman93** [comment]
Just to be complete, this fixes #113 and #120 since those were caused by the target update inherited from q_model

**Comment by AlexKuhnle** [comment]
Thanks @Islandman93 for all the fixes! As we mentioned in Gitter (you have probably seen), we focus on getting out the new version in the next days, so might take a week or so until we integrate this PR.

**Comment by michaelschaarschmidt** [comment]
It seems like this is superseded by the new distributed organisation now?

**Comment by AlexKuhnle** [comment]
I'm just about to change 1-2 things w.r.t. the current architecture using TensorFlow's [Supervisor](https://www.tensorflow.org/api_docs/python/tf/train/Supervisor), since I read that it is or will be [deprecated](https://github.com/tensorflow/tensorflow/issues/6263). The new version will use the [SummarySaverHook](https://www.tensorflow.org/api_docs/python/tf/train/SummarySaverHook) in combination with the [MonitoredSession](https://www.tensorflow.org/api_docs/python/tf/train/MonitoredSession). But I expect that using these classes does automatically handle distributed stuff well (that's what they are there for, I think). So I would suggest we test whether everything is working as expected once it's out.

**Comment by Islandman93** [comment]
Since the api has changed since this PR closing in favor of #214
